Add ignore-*-regex options for table and schema

tomkrouper opened the pull request:
This change will allow users to ignore certain tables or schemas while
using skeema. The main idea behind this is to ignore tables that were
created by `pt-online-schema-change` or `gh-ost`. In addition it makes
sense to allow entire schemas to be ignored besides the system schemas

tomkrouper commented:
Testing of ignoring schemas
<details>
 <summary>The `--ignore-schema-regex` works on init</summary>
```
$ skeema init -h 127.0.0.1 -P 20795 -umsandbox -pmsandbox --ignore-schema-regex="sakila|world_x"
2018-01-03 13:16:52 [INFO]  Creating and using host dir /Users/tomkrouper/tmp/127.0.0.1:20795 for 127.0.0.1:20795

2018-01-03 13:16:52 [INFO]  Populating /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie
2018-01-03 13:16:52 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/event.sql (198 bytes)
2018-01-03 13:16:52 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/pet.sql (257 bytes)

2018-01-03 13:16:52 [INFO]  Skipping schema sakila because of --ignore-schema-regex='sakila|world_x'
2018-01-03 13:16:52 [INFO]  Populating /Users/tomkrouper/tmp/127.0.0.1:20795/world
2018-01-03 13:16:52 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/city.sql (418 bytes)
2018-01-03 13:16:52 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/country.sql (789 bytes)
2018-01-03 13:16:52 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/countrylanguage.sql (437 bytes)

2018-01-03 13:16:52 [INFO]  Skipping schema world_x because of --ignore-schema-regex='sakila|world_x'
```
</details>

tomkrouper commented:
```
skeema push -pmsandbox
2018-01-03 13:44:06 [INFO]  Pushing changes from /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/*.sql to 127.0.0.1:20795 menagerie
2018-01-03 13:44:06 [INFO]  127.0.0.1:20795 menagerie: No differences found

2018-01-03 13:44:06 [INFO]  Pushing changes from /Users/tomkrouper/tmp/127.0.0.1:20795/world/*.sql to 127.0.0.1:20795 world
2018-01-03 13:44:06 [INFO]  127.0.0.1:20795 world: No differences found
```

tomkrouper commented:
`skeema push` also works fine and doesn't care about the missing schemas.
```
skeema push -pmsandbox
2018-01-03 13:44:06 [INFO]  Pushing changes from /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/*.sql to 127.0.0.1:20795 menagerie
2018-01-03 13:44:06 [INFO]  127.0.0.1:20795 menagerie: No differences found

2018-01-03 13:44:06 [INFO]  Pushing changes from /Users/tomkrouper/tmp/127.0.0.1:20795/world/*.sql to 127.0.0.1:20795 world
2018-01-03 13:44:06 [INFO]  127.0.0.1:20795 world: No differences found
```

tomkrouper commented:
Running box `--ignore-schema-regex` and `--ignore-table-regex` results in only the schema(s) being ignored.

```
$ skeema init -h 127.0.0.1 -P 20795 -umsandbox -pmsandbox --ignore-schema-regex="sakila|world_x" --ignore-table-regex="country"
2018-01-03 13:17:53 [INFO]  Creating and using host dir /Users/tomkrouper/tmp/127.0.0.1:20795 for 127.0.0.1:20795

2018-01-03 13:17:53 [INFO]  Populating /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie
2018-01-03 13:17:53 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/event.sql (198 bytes)
2018-01-03 13:17:53 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/menagerie/pet.sql (257 bytes)

2018-01-03 13:17:53 [INFO]  Skipping schema sakila because of --ignore-schema-regex='sakila|world_x'
2018-01-03 13:17:53 [INFO]  Populating /Users/tomkrouper/tmp/127.0.0.1:20795/world
2018-01-03 13:17:53 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/city.sql (418 bytes)
2018-01-03 13:17:53 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/country.sql (789 bytes)
2018-01-03 13:17:53 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/world/countrylanguage.sql (437 bytes)

2018-01-03 13:17:53 [INFO]  Skipping schema world_x because of --ignore-schema-regex='sakila|world_x'
```
TOFIX

tomkrouper commented:
Running a single schema init, the `--ignore-table-regex` is used, but when you run `skeema pull` on the directory, the "ignored" tables are no longer ignored.

```
$ skeema init -h 127.0.0.1 -P 20795 -umsandbox -pmsandbox --schema=world --ignore-table-regex="country"
2018-01-03 13:19:32 [INFO]  Creating and using host dir /Users/tomkrouper/tmp/127.0.0.1:20795 for 127.0.0.1:20795; skipping schema-level subdirs

2018-01-03 13:19:32 [INFO]  Populating /Users/tomkrouper/tmp/127.0.0.1:20795
2018-01-03 13:19:32 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/city.sql (418 bytes)
2018-01-03 13:19:32 [INFO]  Skipping table country because --ignore-table-regex matched country
2018-01-03 13:19:32 [INFO]  Skipping table countrylanguage because --ignore-table-regex matched country
$ cd 127.0.0.1:20705
```
pull
```
$ cd 127.0.0.1:20705
$ skeema pull -pmsandbox
2018-01-03 13:21:01 [INFO]  Updating /Users/tomkrouper/tmp/127.0.0.1:20795 to reflect 127.0.0.1:20795 world
2018-01-03 13:21:01 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/country.sql (789 bytes) -- new table
2018-01-03 13:21:01 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/countrylanguage.sql (437 bytes) -- new table
2018-01-03 13:21:01 [INFO]  Wrote /Users/tomkrouper/tmp/127.0.0.1:20795/city.sql (438 bytes) -- updated file to reflect table alterations
```
TOFIX
(as mentioned above, the city.sql table is updated to include auto-increment.)

tomkrouper commented:
cc: @shlomi-noach

I got past my initial issue of table-regex not working, but there are more issues as seen above. I haven't attempted to work through them yet, but since you asked about this issues I was having, here is where I am so far.

shlomi-noach commented on cmd_init.go:
```diff
@@ -171,6 +180,12 @@ func PopulateSchemaDir(s *tengo.Schema, parentDir *Dir, makeSubdir bool) error {
 		return nil
 	}
+	ignoreSchemaRegex := parentDir.Config.Get("ignore-schema-regex")
+	schemaRE, _ := regexp.Compile(ignoreSchemaRegex)
```
General `golang` advice : you are ignoring the error value returned from `Compile`. This means the user may enter an invalid regex and the app will be silent about it. The advice is to never ignore an `error`. Add something like:
```
schemaRE, err := regexp.Compile(ignoreSchemaRegex)
if err != nil {
  return fmt.Errorf("Invalid regular expression on ignore-schema-regex: %s; %s", ignoreSchemaRegex, err)
}
```

shlomi-noach commented on cmd_init.go:
```diff
@@ -205,7 +220,14 @@ func PopulateSchemaDir(s *tengo.Schema, parentDir *Dir, makeSubdir bool) error {
 	if err != nil {
 		return fmt.Errorf("Cannot obtain table information for %s: %s", s.Name, err)
 	}
+	optionFile, _ := schemaDir.OptionFile()
```
same. Never ignore an `error`

shlomi-noach commented on cmd_init.go:
```diff
@@ -205,7 +220,14 @@ func PopulateSchemaDir(s *tengo.Schema, parentDir *Dir, makeSubdir bool) error {
 		return fmt.Errorf("Cannot obtain table information for %s: %s", s.Name, err)
 	}
+	optionFile, _ := schemaDir.OptionFile()
+	ignoreTableRegex, _ := optionFile.OptionValue("ignore-table-regex")
```
same

shlomi-noach commented on cmd_init.go:
```diff
@@ -205,7 +220,14 @@ func PopulateSchemaDir(s *tengo.Schema, parentDir *Dir, makeSubdir bool) error {
 	}
+	optionFile, _ := schemaDir.OptionFile()
+	ignoreTableRegex, _ := optionFile.OptionValue("ignore-table-regex")
+	re, _ := regexp.Compile(ignoreTableRegex)
```
same

tomkrouper commented:
> ; My lazy guess is that there would be other places throughout the code where schemas should be ignored.
That'll probably work for schema level, but on table level I'm hitting an issue. I see what I need but I don't know how to get to it.

https://github.com/skeema/skeema/blob/c6d61d005184eedad880710290b2e358767c30a5/cmd_pull.go#L112-L113

I `fmt.Printf` out `td` from ^, it prints what's below. I'm not sure how to reference the table name. `td.Statement(mods)` gets the `CreateStatement` from `Table`. It doesn't make sense to me that `td` is showing anything more than that honestly. I'm not sure how to expose the name so I can check the regex.
```
 {%!s(*tengo.Table=&{countrylanguage InnoDB latin1   [0xc420163400 0xc420163480 0xc420163500 0xc420163580] 0xc42014e7e0 [0xc42014e840]  0 true CREATE TABLE `countrylanguage` (
... snip ...
```

tomkrouper commented:
@shlomi-noach do you have any ideas on the previous comment?

shlomi-noach commented:
Sorry, missed the question. I believe the answer lies a little bit later in the same function: `td` is a `TableDiff`, which is an `interface`. In https://github.com/skeema/skeema/blob/c6d61d005184eedad880710290b2e358767c30a5/cmd_pull.go#L117, there is an explicit check that looks at what the specific `struct` it is that implements the interface:
- `CreateTable` - https://github.com/skeema/skeema/blob/c6d61d005184eedad880710290b2e358767c30a5/cmd_pull.go#L118
- `DropTable` - https://github.com/skeema/skeema/blob/c6d61d005184eedad880710290b2e358767c30a5/cmd_pull.go#L134
- `AlterTable` - https://github.com/skeema/skeema/blob/c6d61d005184eedad880710290b2e358767c30a5/cmd_pull.go#L144
- `RenameTable` (not supported yet)

This is the [Type Switch](https://golang.org/doc/effective_go.html#type_switch), and is actually a pretty confusing syntax. The code uses this idiom to cast `td` onto a new variable, confusingly called `td`. It didn't have to be that way, but it was, and it's proper coding but admittedly confusing.

So, anyway, what you'd be looking to do is roughly the same

```go
for _, td := range diff.TableDiffs {
  stmt, err := td.Statement(mods)
  if err != nil {
    return err
  }
  tableName := ""
  switch td := td.(type) {
  case tengo.CreateTable:
    tableName = td.Table.Name
  case tengo.DropTable:
    tableName = td.Table.Name
  case tengo.AlterTable:
    tableName = td.Table.Name
  default:
    return fmt.Errorf("Unsupported diff type %T", td)
  }
  // ... do your thing here

  ... rest of the original code of the function ...
}
```

It looks strange. I'd argue that the `TableDiff` interface should have a `GetTable() * Table` such that we could just `td.GetTable()` without the `switch`.

The above should also explain how you got to see all that fancy info on a `td` objects that seemingly only had `CreateStatement` list. A final tip, while you're `fmt.Printf(...)`-ing, it to use `%+v`. I do that all the time. As example:
```go
   fmt.Printf("Self not that td is %+v \n", td)
```

`%+v` is a special directive in `golang`, that prints out the object along with field names, and is a good representation of field=value for all fields of an object. See https://golang.org/pkg/fmt/#hdr-Printing

evanelias commented:
This is looking very nice, @tomkrouper! Apologies for the delay -- I meant to take a peek a few days ago, but have been a bit slammed with work stuff :)

Just a few quick comments so far:

> `skeema push -pmsandbox`

Quick potential time-saver in manual testing: you can put `password=msandbox` into `~/.my.cnf` and/or one of the `.skeema` files if you want, and Skeema will use that. Skeema just doesn't do that automatically in `skeema init` for security reasons, since it's plaintext.

> Odd realization, skeema pull causes world/city.sql to add the AUTO_INCREMENT value.

Hmm, that is odd. Maybe I have a bug somewhere. The exact behavior is dictated by the [--include-auto-inc](https://github.com/skeema/skeema/blob/master/doc/options.md#include-auto-inc) option, but this defaults to false. And a false value should cause `skeema pull` to only update AUTO_INCREMENT values if they were already present in the table file and had a value > 1.


Regarding getting the table names, @shlomi-noach's advice should do the trick.

> I'd argue that the `TableDiff` interface should have a `GetTable() * Table`  function, so that we could just `td.GetTable()` without the `switch`.

Yes, but it's a bit messy even then. Really `tengo.AlterTable` should probably have two table pointers: the old table and the new table. I can't recall offhand why it doesn't already 🤔 Ditto for `tengo.RenameTable` once it's implemented, especially since I'd imagine you'd want to check the regex against both the old _and_ new names there -- implying that the interface needs the ability to return both the "before" and "after" version of the table.

So that means a few different options for how this would look in the interface:

a) I could add `GetTable() *Table` to the interface, and just document which table it's returning in the case of `AlterTable`. In this case I'm just punting on `RenameTable`'s conceptual issues for now since it's not implemented yet anyway 😅 

b) I could add `Tables() []*Table` which returns a slice of one table (create or drop) or two tables (alter or rename). This solves the rename issue, but feels a little weird.

c) I could add `BeforeTable() *Table` and `AfterTable() *Table` to the interface. This makes the methods clear, but awkward to use: your logic would need to call both, and take special caution because `CreateTable.BeforeTable()` would return nil, and ditto for `DropTable.AfterTable()`.

Bigger picture, I'm not completely happy with the abstractions I've used here around DDL in general. It's full of awkward spots like this. I haven't come up with a better solution yet, but I need to give it more thought. Fingers crossed that I can actually set aside some free time to code this year!

shlomi-noach commented:
> Yes, but it's a bit messy even then. Really tengo.AlterTable should probably have two table pointers: the old table and the new table. I can't recall offhand why it doesn't already 🤔 Ditto for tengo.RenameTable once it's implemented

Oh, true.

tomkrouper commented:
Tested changes with `ignore-table-regex` in both the top level directory and in the schema directory and it worked exactly the same.
```
$ skeema pull -pmsandbox
2018-01-08 15:31:37 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/menagerie to reflect 127.0.0.1:20795 menagerie

2018-01-08 15:31:37 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/world to reflect 127.0.0.1:20795 world
2018-01-08 15:31:37 [INFO]  Skipping table country because --ignore-table-regex matched ^country
2018-01-08 15:31:37 [INFO]  Skipping table countrylanguage because --ignore-table-regex matched ^country

2018-01-08 15:31:37 [INFO]  Skipping schema sakila because of --ignore-schema-regex='sakila|world_x'
2018-01-08 15:31:37 [INFO]  Skipping schema world_x because of --ignore-schema-regex='sakila|world_x'
```

@evanelias thanks for the tips. I'll get the password added in to the file as I move along. I did this intentional to start with because I wanted to provide exact steps. I figured @shlomi-noach would be helping with this some and knew he hadn't played around with it too much and in case he wanted to recreate things I didn't want to jump too far into the weeds. 

I'll take a look at the `City` auto-increment issue after I get the rest of the functionality working. Thanks again for making this tool, and I'm glad I can help to try to make it a little better.

shlomi-noach commented on cmd_push.go:
```diff
@@ -205,13 +206,34 @@ func pushWorker(sps *sharedPushState) {
+			if err != nil {
+				sps.setFatalError(fmt.Errorf("Invalid regular expression on ignore-table-regex: %s; %s", ignoreTableRegex, err))
+				return
+			}
```
This regex creation runs within a loop, and always performs the same operation. It can be extracted outside the loop.
Also, it's the same creation as in https://github.com/skeema/skeema/pull/15/files#diff-1b3966e75764b7531e27b460e0f21b06R181, so you may as well put the regex in a globally accessible place, or otherwise create a single function that creates & returns it.

shlomi-noach replied:
The function, assuming one is written, would return `nil` for the regex.

tomkrouper replied:
The loop it is in assign `t` each time and ignoreTableRegex uses `t`, so I don't think I can move it out.

shlomi-noach replied:
I stand corrected.

shlomi-noach commented on cmd_push.go:
```diff
@@ -205,13 +206,34 @@ func pushWorker(sps *sharedPushState) {
+				default:
+					sps.setFatalError(fmt.Errorf("Unsupported diff type %T", td))
+					return
+				}
```
:+1:

tomkrouper commented:
`skeema pull` is populating `AUTO_INCREMENT=`, but only for some tables. This is truly bizarre. Why some tables and not others. `skeema init` has not added `AUTO_INCREMENT=` at all. Gonna try to dig around and see what is going on. Once this is fixed, I'll go through and run my changes through tests.

tomkrouper commented:
figured out this is something to do with `diff.UnsupportedTables`

evanelias commented:
Ahh, excellent catch! Thank you!

tomkrouper commented:
This is only kind of right, right now. There is no "diff", but the Unsupported tables still re-write the changes because it doesn't go through the same steps as the TableDiffs. Gonna start looking at why so many tables show as unsupported then will swing back to try to do a better fix.

evanelias commented:
Sounds good! The `--debug` option shows more info on why something is marked as unsupported -- it will show a diff of what Skeema expects from information_schema vs what `SHOW CREATE TABLE` actually has.

Often the root cause relates to MySQL version or flavor. Most testing to date has been on mysql 5.6 or percona server 5.6.

tomkrouper commented:
I am indeed using 5.7, so that could be the issue. I'll see what I can figure out. Didn't see anything from adding `--debug`.

<details>
<summary> Output from init & pull trying to figure out unsupported tables</summary>
Not seeing additional info for why something is unsupported
<pre><code>
niffler tmp$ skeema init -h 127.0.0.1 -P 20795 -umsandbox -pmsandbox -d sandbox --debug
2018-01-10 10:54:08 [INFO]  Creating and using host dir /Users/tomkrouper/tmp/sandbox for 127.0.0.1:20795

2018-01-10 10:54:08 [INFO]  Populating /Users/tomkrouper/tmp/sandbox/menagerie
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/menagerie/event.sql (198 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/menagerie/pet.sql (257 bytes)

2018-01-10 10:54:08 [INFO]  Populating /Users/tomkrouper/tmp/sandbox/sakila
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/actor.sql (353 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/address.sql (683 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/category.sql (276 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/city.sql (466 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/country.sql (277 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/customer.sql (866 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/film.sql (1204 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/film_actor.sql (534 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/film_category.sql (575 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/film_text.sql (244 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/inventory.sql (641 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/language.sql (273 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/payment.sql (921 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/rental.sql (990 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/staff.sql (906 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/store.sql (659 bytes)

2018-01-10 10:54:08 [INFO]  Populating /Users/tomkrouper/tmp/sandbox/world
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world/city.sql (418 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world/country.sql (789 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world/countrylanguage.sql (437 bytes)

2018-01-10 10:54:08 [INFO]  Populating /Users/tomkrouper/tmp/sandbox/world_x
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world_x/city.sql (276 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world_x/country.sql (236 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world_x/countryinfo.sql (217 bytes)
2018-01-10 10:54:08 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world_x/countrylanguage.sql (435 bytes)

2018-01-10 10:54:08 [DEBUG] Exit code 0 (SUCCESS)
niffler tmp$ cd sandbox/
menagerie/  sakila/  world/  world_x/
niffler sandbox$ git init
Initialized empty Git repository in /Users/tomkrouper/tmp/sandbox/.git/
niffler sandbox (master #%)$ git add .
niffler sandbox (master +)$ git commit -m "Initial commit"
[master (root-commit) dce320a] Initial commit
 30 files changed, 274 insertions(+)
 create mode 100644 .skeema
 create mode 100644 menagerie/.skeema
 create mode 100644 menagerie/event.sql
 create mode 100644 menagerie/pet.sql
 create mode 100644 sakila/.skeema
 create mode 100644 sakila/actor.sql
 create mode 100644 sakila/address.sql
 create mode 100644 sakila/category.sql
 create mode 100644 sakila/city.sql
 create mode 100644 sakila/country.sql
 create mode 100644 sakila/customer.sql
 create mode 100644 sakila/film.sql
 create mode 100644 sakila/film_actor.sql
 create mode 100644 sakila/film_category.sql
 create mode 100644 sakila/film_text.sql
 create mode 100644 sakila/inventory.sql
 create mode 100644 sakila/language.sql
 create mode 100644 sakila/payment.sql
 create mode 100644 sakila/rental.sql
 create mode 100644 sakila/staff.sql
 create mode 100644 sakila/store.sql
 create mode 100644 world/.skeema
 create mode 100644 world/city.sql
 create mode 100644 world/country.sql
 create mode 100644 world/countrylanguage.sql
 create mode 100644 world_x/.skeema
 create mode 100644 world_x/city.sql
 create mode 100644 world_x/country.sql
 create mode 100644 world_x/countryinfo.sql
 create mode 100644 world_x/countrylanguage.sql
niffler sandbox (master)$ skeema pull -pmsandbox --debug
2018-01-10 10:54:09 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/menagerie to reflect 127.0.0.1:20795 menagerie

2018-01-10 10:54:09 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/sakila to reflect 127.0.0.1:20795 sakila
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/address.sql (683 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/city.sql (466 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/customer.sql (866 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/film.sql (1204 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/inventory.sql (641 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/payment.sql (921 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/rental.sql (990 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/staff.sql (906 bytes) -- updated file to reflect (unsupported) table alterations
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/sakila/store.sql (659 bytes) -- updated file to reflect (unsupported) table alterations

2018-01-10 10:54:09 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/world to reflect 127.0.0.1:20795 world
2018-01-10 10:54:09 [INFO]  Wrote /Users/tomkrouper/tmp/sandbox/world/city.sql (418 bytes) -- updated file to reflect (unsupported) table alterations

2018-01-10 10:54:09 [INFO]  Updating /Users/tomkrouper/tmp/sandbox/world_x to reflect 127.0.0.1:20795 world_x

2018-01-10 10:54:09 [DEBUG] Exit code 0 (SUCCESS)
niffler sandbox (master)$ git diff
niffler sandbox (master)$
</code></pre>
</details>

tomkrouper commented:
I can see why things are supported or not with a `skeema push --debug`.

tomkrouper commented:
Oh skeema, you are so silly.
```
... sample table.
2018-01-10 11:24:25 [WARN]  Table repository_replicas: table uses unsupported features
2018-01-10 11:24:25 [DEBUG] --- Skeema-expected
2018-01-10 11:24:25 [DEBUG] +++ MySQL-actual
2018-01-10 11:24:25 [DEBUG] @@ -13 +13 @@
2018-01-10 11:24:25 [DEBUG] -) ENGINE=InnoDB AUTO_INCREMENT=1288732530 DEFAULT CHARSET=utf8
2018-01-10 11:24:25 [DEBUG] +) ENGINE=InnoDB AUTO_INCREMENT=1288732554 DEFAULT CHARSET=utf8
```

evanelias commented:
Oh! I must have some comparison in there that isn't stripping out the auto-increment values before making the string comparison :(

tomkrouper commented:
> Oh! I must have some comparison in there that isn't stripping out the auto-increment values before making the string comparison :(

Looks like it is here:
https://github.com/skeema/tengo/blob/f7108f2546941565144ac5082d7412ddfcc7a03b/schema.go#L266-L274

Not exactly sure how to deal with this one.

evanelias commented:
Ahh, yes indeed, good find! I think the fix would just involve using `ParseCreateAutoInc` before the comparison there too in https://github.com/skeema/tengo, and then once merged, updating the godep here in https://github.com/skeema/skeema. Timing-wise, I can try a fix this weekend on my end. Or if it's blocking you and you want to take a stab at it, I can hopefully carve out some code review time sooner :) Let me know what you'd prefer, and thanks again for the excellent debugging!

tomkrouper commented:
https://github.com/skeema/tengo/pull/4 worked when I manually updated `vendor/...`, so it looks like it should work.

tomkrouper commented:
> Not exactly sure how to deal with this one.

@evanelias, just so you get a better idea of how I work, I usually will say I have no clue what is going on, then I'll spend time digging in hopes to finding the solution. If you have time to merge the tengo change this week, that would be perfect. Also, not sure how the godep stuff works either.

evanelias commented:
> I usually will say I have no clue what is going on, then I'll spend time digging in hopes to finding the solution

Hahaha, fair enough. Looks like you found the solution very quickly in this case :) Awesome work!

I've merged your PR in tengo, and then updated the vendored dep here in skeema master. Basically this just means I ran `godep update github.com/skeema/tengo` from skeema's base dir, which updates the copy of tengo that is frozen in the vendor subdir. So if you merge upstream master into your branch, it should now work properly without needing a manually hacked `vendor/...`.

tomkrouper commented:
I'm not seeing the change in my `vendor/github.com/skeema/tengo` version. When I ran `godep update github.com/skeema/tengo` to see if that helped, I got this which seemed really odd.
```
niffler skeema (ignore-regex=)$ godep update github.com/skeema/tengo
niffler skeema (ignore-regex *=)$ git status
On branch ignore-regex
Your branch is up to date with 'origin/ignore-regex'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   Godeps/Godeps.json
	deleted:    vendor/github.com/skeema/tengo/LICENSE
	deleted:    vendor/github.com/skeema/tengo/NOTICE
	deleted:    vendor/github.com/skeema/tengo/README.md
	deleted:    vendor/github.com/skeema/tengo/column.go
	deleted:    vendor/github.com/skeema/tengo/ddl.go
	deleted:    vendor/github.com/skeema/tengo/errors.go
	deleted:    vendor/github.com/skeema/tengo/index.go
	deleted:    vendor/github.com/skeema/tengo/instance.go
	deleted:    vendor/github.com/skeema/tengo/schema.go
	deleted:    vendor/github.com/skeema/tengo/table.go
	deleted:    vendor/github.com/skeema/tengo/util.go

no changes added to commit (use "git add" and/or "git commit -a")
```

evanelias commented:
The commit is definitely there in skeema/skeema master :) see bfa1ba91

Just to double-check, have you merged master of skeema/skeema into your ignore-regex branch of tomkrouper/skeema?

tomkrouper commented:
Doh, yeah, I get trying to pull from origin instead of upstream. Thanks.

tomkrouper commented:
Next steps:
* re-test everything (write down tests)
* attempt to break changes
* add examples to docs
* figure out how to merge all commits to one nice commit to send to skeema/skeema

tomkrouper commented:
While testing `skeema diff` I'm seeing the following messages...
```
[mysql] 2018/01/17 14:35:36 packets.go:33: unexpected EOF
[mysql] 2018/01/17 14:39:18 packets.go:33: unexpected EOF
2018-01-17 14:39:18 [ERROR] verifyDiff: Failed to release lock, or connection holding lock already dropped
```
I have no idea if these are things to be worried about.

evanelias commented:
Hmm, those messages are unexpected. For context on what's happening there, Skeema uses mysql's `GET_LOCK` function as a simple way to prevent multiple copies of Skeema from clobbering each other, operating on the same temp schema at the same time. It then calls `RELEASE_LOCK` when done operating on the temp schema. So this error is saying that Skeema attempted to call `RELEASE_LOCK`, but the mysql client connection that previously used `GET_LOCK` already dropped unexpectedly, which means the lock was already released implicitly. The previous error from packets.go also suggests that connection dropping. (packets.go is part of the golang MySQL client driver, http://github.com/go-sql-driver/mysql)

A few questions to help with debugging:
* Does this happen consistently every time you run `skeema diff`, or is it inconsistent?
* How long does a `skeema diff` execution take on your system? If several seconds, are there any databases with an extremely large table count?
* Any aggressive wait_timeout or query killer or anything like that in your env?
* Also maybe anything special going on with MDL lock in your environment? I've read that `GET_LOCK` was changed in 5.7 to use the MDL now, but grasping at straws there :)

If it's a timeout issue, we can manipulate some of the settings using [connect-options](https://github.com/skeema/skeema/blob/master/doc/options.md#connect-options). Failing that, there are some connection pool related things in the driver that we can try.

Side note: I may remove/change the locking stuff some day soon anyway, since it's not really a comprehensive solution. Really Skeema needs a better notion of persistent state anyway, since some alters may take a long time to run, and subsequent executions of Skeema should be able to understand an alter is already "in-flight" for a given table. Currently there isn't support for that yet.

tomkrouper commented:
Well, I just attempted to squash all my commits into one and I don't seem to have managed that. :-/

tomkrouper commented:
> Does this happen consistently every time you run skeema diff, or is it inconsistent?

It happens every time on busier clusters. I ran it late at night on a busy cluster and it happened less, but still happened.

> How long does a skeema diff execution take on your system? If several seconds, are there any databases with an extremely large table count?

The cluster I'm testing on has 382 tables.

> Any aggressive wait_timeout or query killer or anything like that in your env?

I'm running as myself, so pt-kill ignores me. 30 second wait_timeout on the host I'm hitting.

> Also maybe anything special going on with MDL lock in your environment? I've read that GET_LOCK was changed in 5.7 to use the MDL now, but grasping at straws there :)

I have not checked on this but I will.

tomkrouper commented:
> figure out how to merge all commits to one nice commit to send to skeema/skeema

merge and squash will take care of this.

tomkrouper commented:
> re-test everything (write down tests)

```
skeema init -h 127.0.0.1 -P 20795 -umsandbox -pmsandbox --ignore-schema-regex="sakila|world_x" --ignore-table-regex="^country" -d sandbox --debug
skeema pull [--ignore-schema-regex=...]  [--ignore-table-regex=...]
skeema diff [--ignore-schema-regex=...]  [--ignore-table-regex=...]
skeema push [--ignore-schema-regex=...]  [--ignore-table-regex=...]
# Modifying sql files, modifying database and check that everything matched as it should.
# optionally adding ignores to test how things are handled. Everything ran as expected.
```

> attempt to break changes

The closest thing to a breakage I got was some complaints from the go mysql driver. Though I got messages, it still seems that skeema acted as it would be expected.

evanelias commented:
From visual inspection, this is looking great! I'll aim to do a more thorough review tonight, but so far so good.

My only quick nitpicky requests so far:
* If ready for full review (sounds like it?) please remove the WIP from the PR title ;)
* Have you tested against `skeema lint`? Ideally these options should still work there too, skipping those schemas/tables for linting purposes.
* How do you feel about shorter option names, just for easier typing? I was thinking maybe drop the "-regex" suffix, so the options are just called `ignore-schema` and `ignore-table`. As long as the docs and help text indicate they're regexes, I think it's still pretty clear. (Side note: at some point soon, I'll add support to the [mybase cli framework](https://github.com/skeema/mybase) for a regex option type, which should remove some of the boilerplate around regex compilation and validation.)

tomkrouper commented:
> Have you tested against skeema lint? Ideally these options should still work there too, skipping those schemas/tables for linting purposes.

If you manually create a table that is part of `ignore-table`, `skeema lint` will still lint the file. It however will not show up in diff, pull, or push. I'll get lint to complain too.

tomkrouper commented:
Several updates, the commits explain the changes. `skeema lint` now works as it should.

evanelias commented:
Everything looks perfect from my testing. Thanks again for this excellent work @tomkrouper! Shall I go ahead and do squash & merge?

tomkrouper commented:
At your convenience. Thanks again for creating this tool. Glad I could help out a little. 🙇
